src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -7320,6 +7342,69 @@ private struct NetResource
             [MarshalAs(UnmanagedType.LPWStr)]
             public string Provider;
         }
+
+        #region InodeTracker
+        /// <summary>
+        /// Tracks visited files/directories by caching their device IDs and inodes.
+        /// </summary>
+        private class InodeTracker
+        {
+            private HashSet<(UInt64, UInt64)> _visitations;
+
+            /// <summary>
+            /// Construct a new InodeTracker with an initial path
+            /// </summary>
+            internal InodeTracker(string path)
+            {
+                _visitations = new HashSet<(UInt64, UInt64)>();
+
+                if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out (UInt64, UInt64) inodeData))
+                {
+                    _visitations.Add(inodeData);
+                }
+            }
+
+            /// <summary>
+            /// Determine if a path has been visited. Optionally mark the path as
+            /// having been visited if it has not already
+            /// </summary>
+            /// <param name="path">
+            /// Path to the file system item to be checked.
+            /// </param>
+            /// <param name="visitIfNotAlready">
+            /// Flag to indicate whether the path should be marked as visited if
+            /// it has not already been visited.
```

Comment:
I give this some more thinking, and believes that the `TryVisitPath` signature is more natural in this case. But we need to follow the `Try<>` pattern.

Again, take `HashSet.Add(object)` as an example:
1. If the object is not in the set, add it and return true.
2. If the object is already in the set, return false which means it was not added in this call because it's already there.

Similarly, `InodeTracker.TryVisitPath(path)` should be
1. If the path is not visited yet, visit it and return true.
2. If the path is already visited, return false which means it was not visited in this call because it's already visited.

The code would be like:
```CSharp
internal bool TryVisitPath(string path)
{
    bool returnValue = false;
    if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out (UInt64, UInt64) inodeData))
    {
        returnValue = _visitations.Add(inodeData);
    }
    return returnValue;
}
```

@jeffbi @iSazonov what do you think?